File: backend/src/main/kotlin/Firebase.kt
package com.scienta

import com.google.auth.oauth2.GoogleCredentials
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.cloud.FirestoreClient
import com.google.cloud.firestore.Firestore
import java.io.ByteArrayInputStream

fun initFirebase(config: AppConfig) {
    if (FirebaseApp.getApps().isNotEmpty()) return

    val serviceAccount = ByteArrayInputStream(config.firebaseServiceAccountJson.toByteArray())
    val credentials = GoogleCredentials.fromStream(serviceAccount)

    val options = FirebaseOptions.builder()
        .setCredentials(credentials)
        .setProjectId(config.firebaseProjectId)
        .build()

    FirebaseApp.initializeApp(options)
}

fun getFirestore(): Firestore = FirestoreClient.getFirestore()

fun getFirebaseAuth(): FirebaseAuth = FirebaseAuth.getInstance()

File: frontend/src/firebase.ts
import { initializeApp, type FirebaseApp } from 'firebase/app'
import { getAuth, type Auth } from 'firebase/auth'
import { getFirestore, type Firestore } from 'firebase/firestore'

let app: FirebaseApp
let auth: Auth
let db: Firestore

export async function initFirebase() {
  const res = await fetch('/config')
  const config = await res.json()
  app = initializeApp(config)
  auth = getAuth(app)
  db = getFirestore(app)
}

export { auth, db }
